gnucash/import-export/import-main-matcher.c

```diff
@@ -885,26 +894,118 @@ gnc_gen_trans_assign_transfer_account_to_selection_cb (GtkMenuItem *menuitem,
     LEAVE("");
 }
 
-typedef enum
+typedef struct
+{
+    Split *split;
+    Transaction *trans;
+    GtkTreeIter iter;
+    const char *orig_desc, *orig_notes, *orig_memo;
+} RowInfo;
+
+static RowInfo * row_get_info (gpointer row, GNCImportMainMatcher *info)
+{
+    GNCImportTransInfo *trans_info;
+    GtkTreeModel *model = gtk_tree_view_get_model (info->view);
+    RowInfo *retval = g_new (RowInfo, 1);
+    gtk_tree_model_get_iter (model, &retval->iter, row);
+    gtk_tree_model_get (model, &retval->iter,
+                        DOWNLOADED_COL_DATA, &trans_info,
+                        DOWNLOADED_COL_DESCRIPTION_ORIGINAL, &retval->orig_desc,
+                        DOWNLOADED_COL_NOTES_ORIGINAL, &retval->orig_notes,
+                        DOWNLOADED_COL_MEMO_ORIGINAL, &retval->orig_memo,
+                        -1);
+    retval->trans = gnc_import_TransInfo_get_trans (trans_info);
+    retval->split = gnc_import_TransInfo_get_fsplit (trans_info);
+    return retval;
+}
+
+static void populate_list (gpointer key, gpointer value, GtkListStore *list)
+{
+    GtkTreeIter iter;
+    gtk_list_store_append (list, &iter);
+    gtk_list_store_set (list, &iter, 0, key, 1, g_utf8_collate_key (key, -1), -1);
+}
+
+static void
+setup_entry (GtkWidget *entry, gboolean sensitive, GHashTable *hash,
+             const char *initial)
+{
+    GtkEntryCompletion* completion;
+    GtkListStore *list;
+
+    gtk_entry_set_text (GTK_ENTRY (entry), sensitive ? initial : _("Disabled"));
+    gtk_widget_set_sensitive (entry, sensitive);
+    if (!sensitive)
+        return;
+
+    g_hash_table_insert (hash, (gpointer)initial, GINT_TO_POINTER (1));
+    list = gtk_list_store_new (2, G_TYPE_STRING, G_TYPE_STRING);
+    g_hash_table_foreach (hash, (GHFunc)populate_list, list);
+    gtk_tree_sortable_set_sort_column_id (GTK_TREE_SORTABLE (list), 1,
+                                          GTK_SORT_ASCENDING);
+
+    completion = gtk_entry_completion_new ();
+    gtk_entry_completion_set_model (completion, GTK_TREE_MODEL(list));
+    gtk_entry_completion_set_text_column (completion, 0);
+    gtk_entry_set_completion (GTK_ENTRY (entry), completion);
+}
+
+static gboolean
+input_new_fields (GtkWidget *parent, RowInfo *info, GtkTreeStore *store,
+                  gboolean edit_desc, gboolean edit_notes, gboolean edit_memo,
+                  GHashTable *desc_hash, GHashTable *notes_hash, GHashTable *memo_hash,
+                  char **new_desc, char **new_notes, char **new_memo)
```

Comment:
Four of the parameters at the call site come from the GncMainMatcher* so maybe just take that as a parameter to cut down the size of the parameter list; wrapping the hash tables in a struct (see comment about creating that once) will help too.